tensorboard/plugins/hparams/summary_v2_test.py

```diff
@@ -43,6 +43,164 @@
+    self.hparams = {
+        hp.HParam("learning_rate", hp.RealInterval(1e-2, 1e-1)): 0.02,
+        hp.HParam("dense_layers", hp.IntInterval(2, 7)): 5,
+        hp.HParam("optimizer", hp.Discrete(["adam", "sgd"])): "adam",
+        hp.HParam("who_knows_what"): "???",
+        hp.HParam(
+            "magic",
+            hp.Discrete([False, True]),
+            display_name="~*~ Magic ~*~",
+            description="descriptive",
+        ): True,
+        "dropout": 0.3,
+    }
+    self.normalized_hparams = {
+        "learning_rate": 0.02,
+        "dense_layers": 5,
+        "optimizer": "adam",
+        "who_knows_what": "???",
+        "magic": True,
+        "dropout": 0.3,
+    }
+    self.start_time_secs = 123.45
+    self.group_name = "big_sha"
+
+    self.expected_session_start_pb = plugin_data_pb2.SessionStartInfo()
+    text_format.Merge(
+        """
+        hparams { key: "learning_rate" value { number_value: 0.02 } }
+        hparams { key: "dense_layers" value { number_value: 5 } }
+        hparams { key: "optimizer" value { string_value: "adam" } }
+        hparams { key: "who_knows_what" value { string_value: "???" } }
+        hparams { key: "magic" value { bool_value: true } }
+        hparams { key: "dropout" value { number_value: 0.3 } }
+        group_name: "big_sha"  # we'll ignore this field when asserting equality
+        """,
+        self.expected_session_start_pb,
+    )
+    self.expected_session_start_pb.start_time_secs = self.start_time_secs
+
+  def _check_summary(self, summary_pb):
+    """Test that a summary contains exactly the expected hparams PB."""
+    values = summary_pb.value
+    self.assertEqual(len(values), 1, values)
+    actual_value = values[0]
+    self.assertEqual(
+        actual_value.metadata.plugin_data.plugin_name,
+        metadata.PLUGIN_NAME,
+    )
+    plugin_content = actual_value.metadata.plugin_data.content
+    info_pb = metadata.parse_session_start_info_plugin_data(plugin_content)
+    # Ignore the `group_name` field; its properties are checked separately.
+    info_pb.group_name = self.expected_session_start_pb.group_name
+    self.assertEqual(info_pb, self.expected_session_start_pb)
+
+  def _check_logdir(self, logdir):
+    """Test that the hparams summary was written to `logdir`."""
+    self._check_summary(_get_unique_summary(self, logdir))
+
+  def test_eager(self):
+    with tf.compat.v2.summary.create_file_writer(self.logdir).as_default():
+      result = hp.hparams(self.hparams, start_time_secs=self.start_time_secs)
+      self.assertTrue(result)
+    self._check_logdir(self.logdir)
+
+  def test_graph_mode(self):
+    with \
+        tf.compat.v1.Graph().as_default(), \
+        tf.compat.v1.Session() as sess, \
+        tf.compat.v2.summary.create_file_writer(self.logdir).as_default() as w:
+      sess.run(w.init())
+      summ = hp.hparams(self.hparams, start_time_secs=self.start_time_secs)
+      self.assertTrue(sess.run(summ))
```

Comment:
You probably want `sess.run(w.flush())` after running the summary ops.  Without it I would expect this test to be flaky.

Or you can set `max_queue` on the writer to 0.

Reply:
I thought that exiting the context manager would flush the writer
automatically. How is this context different from
<https://github.com/tensorflow/tensorboard/pull/2139#discussion_r278677673>?
Even though this one is in legacy graph mode, both use the `.as_default`
method on a V2 summary writer object, right?

Reply:
Hmm—from a bit of experimentation, it is _not_ guaranteed to flush when
the writer context manager exits, but _does_ always seem to flush when
the session context manager exits. That’s sufficiently non-obvious that
I’ll go ahead and add the explicit `flush()`, though I would still be
interested to understand why this differs from the situation linked in
my previous comment.